data/scenarios/Challenges/_bucket-brigade/brigade.sw

```diff
@@ -0,0 +1,75 @@
+def doN = \n. \f. if (n > 0) {f; doN (n - 1) f} {}; end;
+
+def pollUntilHas = \item.
+    itemInInventory <- has item;
+    if itemInInventory {} {
+        pollUntilHas item;
+    };
+    end;
+
+def pollUntilCount = \targetCount. \item.
+    itemCount <- count item;
+    if (itemCount < targetCount) {
+        pollUntilCount targetCount item;
+    } {};
+    end;
+
+def waitToGiveUntilHas = \item. \recipient.
+    pollUntilHas item;
+    give recipient item;
+    end;
+
+/** The "depth" is the number of spaces between
+the root robot and the target item.
+It is one minus the distance to the target item.
+*/
+def recurseUntilDepth = \depth.
+    let item = "coal" in
+    child <- build {
+
+        // NOTE: Treads are auto-installed via the implicit "require"
```

Comment:
There are two things that made this solution tricky to implement.

1. The implicit `require` behavior steals things out of my inventory and equips them on the robot being built.  I'm not sure the benefit of such behavior outweighs the "surprise".  At least, perhaps, we should be able to disable that behavior?
2. Due to the above behavior, my `build` command was failing.  I narrowed the failure of my program down to the `build` command via `printf`-style debugging with the `say` command.  However, even then I was not sure **why** `build` was failing.  One cannot use the `try` command to retrieve the error message (and `say` it) of the actual failure that the `try` caught.  I considered using `upload`, assuming that the error message is part of the log.  But the error itself caused my child robot's program to halt, precluding an `upload` step after encountering the error.

In the end I cheated by using the web API to find out what the actual error was:

        curl -s http://localhost:5357/robots | jq -C '.[]._robotLog[] | select( ._leSaid == "ErrorTrace" )' | less -R

It told me that the `build` command failed because I didn't have `treads` in my inventory (and this was because they were `equipped`).  So the solution in this case was to remember to `unequip` the treads right after using them to `move`.

Reply:
Yes, the implicit `require` behavior of `build` is intended to make things easier in most situations, but it could make sense to have some sort of `raw_build` command which does not do any requirements analysis for you.

I debugged my solution by starting with the `--cheat` flag, then when something went wrong I switched into Creative mode so I could build a robot to go `salvage` all the dead robots and bring me back their logs.  That worked pretty well.

Reply:
#921 921